src/Compression/CompressionCodecQATZSTD.cpp

```diff
@@ -0,0 +1,114 @@
+        /* Start QAT device, start QAT device at any time before compression job started */
+        int res = QZSTD_startQatDevice();
+        /* Create sequence producer state for QAT sequence producer */
+        sequenceProducerState = QZSTD_createSeqProdState();
+        /* register qatSequenceProducer */
+        ZSTD_registerSequenceProducer(
+            cctx,
+            sequenceProducerState,
+            qatSequenceProducer
+        );
+        /* Enable sequence producer fallback */
+        ZSTD_CCtx_setParameter(cctx, ZSTD_c_enableSeqProducerFallback, 1);
+        ZSTD_CCtx_setParameter(cctx, ZSTD_c_compressionLevel, level);
+        initialized = true;
+        LOG_WARNING(log, "Initialization of hardware-assisted(QAT) ZSTD codec result: {} ", static_cast<UInt32>(res));
+    }
+    size_t compressed_size = ZSTD_compress2(cctx, dest, ZSTD_compressBound(source_size), source, source_size);
+
+    if (ZSTD_isError(compressed_size))
+        throw Exception(ErrorCodes::CANNOT_COMPRESS, "Cannot compress with ZSTD codec: {}", std::string(ZSTD_getErrorName(compressed_size)));
+
+    return static_cast<UInt32>(compressed_size);
+}
+
+void registerCodecQATZSTD(CompressionCodecFactory & factory)
+{
+    factory.registerCompressionCodec("QATZSTD", {}, [&](const ASTPtr & arguments) -> CompressionCodecPtr
+    {
+        int level = CompressionCodecZSTD::ZSTD_DEFAULT_LEVEL;
+        if (arguments && !arguments->children.empty())
+        {
+            if (arguments->children.size() > 1)
+                throw Exception(ErrorCodes::ILLEGAL_SYNTAX_FOR_CODEC_TYPE, "QATZSTD codec must have 1 parameter, given {}", arguments->children.size());
+
+            const auto children = arguments->children;
+            const auto * literal = children[0]->as<ASTLiteral>();
+            if (!literal)
+                throw Exception(ErrorCodes::ILLEGAL_CODEC_PARAMETER, "QATZSTD codec argument must be integer");
+
+            level = static_cast<int>(literal->value.safeGet<UInt64>());
+            if (level > CompressionCodecQATZSTD::QATZSTD_SUPPORTED_MAX_LEVEL || level < CompressionCodecQATZSTD::QATZSTD_SUPPORTED_MIN_LEVEL)
+                throw Exception(ErrorCodes::ILLEGAL_CODEC_PARAMETER,
+                    "QATZSTD codec doesn't support level more than {} and lower than {} , given {}",
+                    CompressionCodecQATZSTD::QATZSTD_SUPPORTED_MAX_LEVEL, CompressionCodecQATZSTD::QATZSTD_SUPPORTED_MIN_LEVEL, level);
+        }
+
+        return std::make_shared<CompressionCodecQATZSTD>(level);
+    });
+}
+
+CompressionCodecQATZSTD::CompressionCodecQATZSTD(int level_)
+    : CompressionCodecZSTD(level_), level(level_), initialized(false), log(&Poco::Logger::get("CompressionCodecQATZSTD"))
+{
+    setCodecDescription("QATZSTD", {std::make_shared<ASTLiteral>(static_cast<UInt64>(level))});
+}
+
+CompressionCodecQATZSTD::~CompressionCodecQATZSTD()
+{
+    if (initialized)
+    {
+        /* Free sequence producer state */
```

Comment:
(Cosmetic: ClickHouse uses `/// ` for comments)

`ZSTD_freeCCtx` returns an error code, I guess we should check it and write a warning to the log if something is unexpected.

Reply:
Done